mm2src/trezor/src/eth/eth_command.rs

```diff
@@ -0,0 +1,213 @@
+        show_display: bool,
+    ) -> TrezorResult<TrezorResponse<'a, 'b, XPub>> {
+        let req = proto_ethereum::EthereumGetPublicKey {
+            address_n: serialize_derivation_path(derivation_path),
+            show_display: Some(show_display),
+        };
+        let result_handler = ResultHandler::new(|m: proto_ethereum::EthereumPublicKey| Ok(m.xpub));
+        self.call(req, result_handler).await
+    }
+
+    pub async fn sign_eth_tx(
+        &mut self,
+        derivation_path: &DerivationPath,
+        unsigned_tx: &UnSignedEthTx,
+        chain_id: u64,
+    ) -> TrezorResult<UnverifiedEthTx> {
+        let mut data: Vec<u8> = vec![];
+        let req = to_sign_eth_message(unsigned_tx, derivation_path, chain_id, &mut data);
+        let processor = self
+            .processor
+            .as_ref()
+            .or_mm_err(|| TrezorError::InternalNoProcessor)?
+            .clone();
+        let mut tx_request = self
+            .send_sign_eth_tx(req)
+            .await?
+            .process(processor.clone())
+            .await
+            .mm_err(|e| TrezorError::Internal(e.to_string()))?;
+
+        while let Some(data_length) = tx_request.data_length {
+            if data_length > 0 {
+                let req = proto_ethereum::EthereumTxAck {
+                    data_chunk: data.splice(..data_length as usize, []).collect(),
+                };
+                tx_request = self
+                    .send_eth_tx_ack(req)
+                    .await?
+                    .process(processor.clone())
+                    .await
+                    .mm_err(|e| TrezorError::Internal(e.to_string()))?;
+            } else {
+                break;
+            }
+        }
+
+        let sig = extract_eth_signature(&tx_request)?;
+        Ok(unsigned_tx.clone().with_signature(sig, Some(chain_id)))
+    }
+
+    async fn send_sign_eth_tx<'b>(
+        &'b mut self,
+        req: proto_ethereum::EthereumSignTx,
+    ) -> TrezorResult<TrezorResponse<'a, 'b, proto_ethereum::EthereumTxRequest>> {
+        let result_handler = ResultHandler::<proto_ethereum::EthereumTxRequest>::new(Ok);
+        self.call(req, result_handler).await
+    }
+
+    async fn send_eth_tx_ack<'b>(
+        &'b mut self,
+        req: proto_ethereum::EthereumTxAck,
+    ) -> TrezorResult<TrezorResponse<'a, 'b, proto_ethereum::EthereumTxRequest>> {
+        let result_handler = ResultHandler::<proto_ethereum::EthereumTxRequest>::new(Ok);
+        self.call(req, result_handler).await
+    }
+}
+
+fn to_sign_eth_message(
+    unsigned_tx: &UnSignedEthTx,
+    derivation_path: &DerivationPath,
+    chain_id: u64,
+    data: &mut Vec<u8>,
+) -> proto_ethereum::EthereumSignTx {
+    // if we have it, pass network or token definition info to show on the device screen:
+    let eth_defs = proto_ethereum_definitions::EthereumDefinitions {
+        encoded_network: get_eth_network_def(chain_id),
+        encoded_token: None, // TODO add looking for tokens defs
```

Comment:
Can you please explain this todo in more details?

Reply:
they support a list of tokens embedded into the firmware (this allows to show token name on the display, when confirming).
We can put additional tokens here (but only if they have their definitions on their website, bcz definitions are signed with trezor key)